File: minSubarrayLen.py
'''
Approach- Sliding Window 

'''
def minSubarrayLen(target,nums): # TC O(2n) --> TC O(n) // SC O(1)
	sum,l=0,0
	ans=float('inf')
	for r in range(len(nums)):
		sum+=nums[r]
		while sum>=target:
			ans=min(ans,r-l+1)
			sum-=nums[l]
			l+=1
	return 0 if ans==float('inf') else ans 

print(minSubarrayLen(7,[2,3,1,2,4,3]))

File: minSubArrayLen.java
import java.util.*;
// Approach 1- Sliding window
class Solution { // TC O(n) SC O(1)
    public int minSubArrayLen(int target, int[] nums) {
        int left=0, right=0;
        int ans=Integer.MAX_VALUE;
        int sum=0;
        while(right<nums.length){
            sum+=nums[right];
            
            while(sum>=target){
                ans=Math.min(ans, right-left+1);
                sum-=nums[left];
                left++;
            }
            right++;
        }
        return ans==Integer.MAX_VALUE ? 0: ans;
    }
}

class minSubArrayLen{
    public static void main(String[] args) {
        int[] nums={2,3,1,2,4,3};
        int target=7;
        Solution s= new Solution();
        System.out.println(s.minSubArrayLen(target, nums));

    }
}